File: C++/k_closest_to_origin.cpp
class Solution {
public:
    typedef pair<int, int> pp;
    // Using priority method using min heap
    // Time complexity: O(n)
    // Space complexity: O(n)
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        vector<vector<int>> ret;
        priority_queue<pp, vector<pp>, greater<pp>> pq;
        for (int i = 0; i < points.size(); ++i) {
            int distance = points[i][0] * points[i][0] + points[i][1] * points[i][1];
            pq.push(make_pair(distance, i));
        }
        for (int i = 0; i < k; ++i) {
            auto top = pq.top();
            ret.push_back(points[top.second]);
            pq.pop();
        }
        return ret;
    }
    
    // Using priority method using max heap
    // Time complexity: O(n)
    // Space complexity: O(k)
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        vector<vector<int>> ret;
        priority_queue<pp> pq;
        for (int i = 0; i < points.size(); ++i) {
            int distance = points[i][0] * points[i][0] + points[i][1] * points[i][1];
            if(i < k) {
                pq.push(make_pair(distance, i));
            }
            else if (distance < pq.top().first){
                pq.pop();
                pq.push(make_pair(distance, i));
            }
        }
        for (int i = 0; i < k; ++i) {
            auto top = pq.top();
            ret.push_back(points[top.second]);
            pq.pop();
        }
        return ret;
    }
    
    int distance(vector<int> p) {
        return p[0] * p[0] + p[1] * p[1];
    }
    
    int partition(vector<vector<int>>& points, int left, int right) {
        int mid = (left + right) / 2;
        int index = left;
        swap(points[mid], points[left]);
        for (int i = left+1; i < right+1; ++i) {
            if (distance(points[i]) <= distance(points[left])) {
                index++;
                swap(points[i], points[index]);
            }
        }
        if (index < right + 1) {
            swap(points[index], points[left]);
        }
        
        return index;
    }
    // Using partition method
    // Time complexity: O(n)
    // Space complexity: O(k)
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        int l = points.size();
        int left = 0;
        int right = l-1;
        int pivot = 0;
        vector<vector<int>> ret;
        while (pivot != k && left <= right) {
            pivot = partition(points, left, right);
            if (pivot < k) {
                left = pivot + 1;
            }
            else if (pivot > k) {
                right = pivot - 1;
            }
        }
        for (int i = 0; i < k ; ++i) {
            ret.push_back(points[i]);
        }
        return ret;
    }
};

File: Python/k_closest_to_origin.py
import heapq
class Solution(object):
    def __init__(self):
        self.name = 'S->13'

    # Using sorting method:
    # Time complexity: O(nlogn)
    # Space complexity: O(logn) to O(n) in worst case
    def kClosestA(self, points, k):
        points.sort(key=lambda x: x[0] ** 2 + x[1] ** 2)
        return points[:k]

    # Using min heap method
    # Time complexity: O(n)
    # Space complexity: O(n)
    def kClosestB(self, points, k):
        heap = []
        ret = []
        for p in points:
            distance = p[0] ** 2 + p[1] ** 2
            heapq.heappush(heap, [distance, p[0], p[1]])
        for _ in range(k):
            smallest = heapq.heappop(heap)
            ret.append([smallest[1], smallest[2]])
        return ret

    def __distance(self, p):
        return p[0] ** p[0] + p[1] ** p[1]

    # def __patition(self, points, left, right):
    #     mid = (left + right) // 2
    #     points[left], points[mid] = points[mid], points[left]
    #     index = left
    #     for i in range(left+1, right+1):
    #         if (self.__distance(points[i]) <= self.__distance(points[left])):
    #             index += 1
    #             points[i], points[index] = points[index], points[i]
    #     if index < right + 1:
    #         points[index], points[left] = points[left], points[index]
    #     return index

    # def __patition(self, points, left, right):
    #     mid = (left + right) // 2
    #     pivot = self.__distance(points[mid])
    #     while left < right:
    #         while left <= right and self.__distance(points[left]) < pivot:
    #             left += 1
    #         while self.__distance(points[right]) > pivot:
    #             right -= 1
    #         if left < right:
    #             points[left], points[right] = points[right], points[left]
    #             left += 1
    #             right -= 1
    #     return left

    def __patition(self, points, left, right):
        mid = (left + right) // 2
        pivot = self.__distance(points[mid])
        while left < right:
            if self.__distance(points[left]) >= pivot:
                points[left], points[right] = points[right], points[left]
                right -= 1
            else:
                left += 1
        if self.__distance(points[left]) < pivot:
            left += 1
        return left

    def kClosestC(self, points, k):
        left = 0
        right = len(points) - 1
        pivot = len(points)
        while pivot != k:
            pivot = self.__patition(points, left, right)
            if pivot < k:
                left = pivot
            elif pivot > k:
                right = pivot - 1

        return points[:k]


# points = [[1, 1], [3, 3], [2, 2], [4, 4], [-1, -1]]
# k = 2
# print(Solution().kClosestA(points, k))
# print(Solution().kClosestB(points, k))
print(Solution().kClosestC([[0,1],[1,0]], 2))
# print(Solution().kClosestC([[1, 1], [3, 3], [2, 2], [4, 4], [-1, -1]], 2))
# print(Solution().kClosestC([[1,3],[-2,2]], 1))
# print(Solution().kClosestC([[1,3],[-2,2],[2,-2]], 2))
